integration_test.go

```diff
@@ -67,36 +62,90 @@ func runsOnEmulator() bool {
 }
 
-func initEmulator() error {
+func initTestInstance(config string) (cleanup func(), err error) {
 	ctx := context.Background()
 	instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	defer instanceAdmin.Close()
+	// Check if the instance exists or not.
+	_, err = instanceAdmin.GetInstance(ctx, &instancepb.GetInstanceRequest{
+		Name: fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId),
+	})
+	if err == nil {
+		return func() {}, nil
+	}
+	if spanner.ErrCode(err) != codes.NotFound {
+		return nil, err
+	}
+
+	// Instance does not exist. Create a temporary instance for this test run.
+	// The instance will be deleted after the test run.
 	op, err := instanceAdmin.CreateInstance(ctx, &instancepb.CreateInstanceRequest{
 		Parent:     fmt.Sprintf("projects/%s", projectId),
 		InstanceId: instanceId,
 		Instance: &instancepb.Instance{
-			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, "emulator-config"),
+			Config:      fmt.Sprintf("projects/%s/instanceConfigs/%s", projectId, config),
 			DisplayName: instanceId,
 			NodeCount:   1,
+			Labels: map[string]string{
+				"gosqltestinstance": "true",
+				"createdat":         fmt.Sprintf("t%d", time.Now().Unix()),
+			},
 		},
 	})
 	if err != nil {
-		// Check if the instance has already been created by another (parallel) test.
-		code := spanner.ErrCode(err)
-		if code != codes.AlreadyExists {
-			return fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
-		}
+		return nil, fmt.Errorf("could not create instance %s: %v", fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId), err)
 	} else {
 		// Wait for the instance creation to finish.
 		_, err := op.Wait(ctx)
 		if err != nil {
-			return fmt.Errorf("waiting for instance creation to finish failed: %v", err)
+			return nil, fmt.Errorf("waiting for instance creation to finish failed: %v", err)
 		}
 	}
-	return nil
+	// Delete the instance after all tests have finished.
+	// Also delete any stale test instances that might still be around on the project.
+	return func() {
+		instanceAdmin, err := instance.NewInstanceAdminClient(ctx)
+		if err != nil {
+			return
+		}
+		// Delete this test instance.
+		instanceAdmin.DeleteInstance(ctx, &instancepb.DeleteInstanceRequest{
+			Name: fmt.Sprintf("projects/%s/instances/%s", projectId, instanceId),
+		})
+		// Also delete any other stale test instance.
+		instances := instanceAdmin.ListInstances(ctx, &instancepb.ListInstancesRequest{
+			Parent: fmt.Sprintf("projects/%s", projectId),
+			Filter: "label.gosqltestinstance:*",
+		})
+		for {
+			instance, err := instances.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				log.Printf("failed to fetch instances during cleanup: %v", err)
+				break
+			}
+			if createdAtString, ok := instance.Labels["createdat"]; ok {
+				// Strip the leading 't' from the value.
```

Comment:
It seems odd to include the 't' if we are just going to strip it again later.

Reply:
Label values must start with a letter: https://cloud.google.com/spanner/docs/reference/rpc/google.spanner.admin.instance.v1#instance

Reply:
Oh interesting. I think that documentation may be out of date because we use the raw numbers in Python:
https://github.com/googleapis/python-spanner/blob/master/tests/system/conftest.py#L117-L118

Reply:
That's interesting. I think that it might actually work on product, but it certainly failed on the emulator (I tried without the 't' first). But I guess that would be logical if it is a recent change to production.